src/interaction/VerletListInteractionTemplate.hpp

```diff
@@ -118,12 +118,15 @@ namespace espressopp {
     addForces() {
       LOG4ESPP_DEBUG(_Potential::theLogger, "loop over verlet list pairs and add forces");
 
+      int vlmaxtype = verletList->getMaxType();
+      Potential max_pot = potentialArray.at(vlmaxtype,vlmaxtype); // force a resize
```

Comment:
I used the following [python script](https://gist.githubusercontent.com/jnvance/5b04c6abc0c5ac899bfa4c4639d94efd/raw/6ddf43c193e6eb773c1c4d8671554e06641b4c45/test.py) with this [input file](https://gist.githubusercontent.com/jnvance/5b04c6abc0c5ac899bfa4c4639d94efd/raw/6ddf43c193e6eb773c1c4d8671554e06641b4c45/lennard_jones_fluid_10000_10000.xyz).

I compiled Espresso++ with Intel Compiler 2018.1 with `-DCMAKE_CXX_FLAGS="-g -DLOG4ESPP_LEVEL_ERROR -O3 -xSKYLAKE-AVX512 -qopt-zmm-usage=high"` and ran the simulations on a single core of Intel Gold 6130 (Skylake) on Mogon II.

I think determining `max_type` at every rebuild allows for the rare case that particles types that exist only in one subdomain can have an entry in the potentialArray in case they move to another subdomain.

For example if a particle of type 2 exists in node A but not  in node B, and does not have an entry in potentialArray, when that particle moves to node B, the only way for potentialArray in node B to have an entry for type 2 is to force a resize. In the current master, this is done when `getPotential` is called in the force calculation loop. Since I removed this from the loop, it has to be done before the loop starts, otherwise a segmentation fault could occur. If there is no change to max_type, then there would be no resize, and the single range check shouldn't cost that much.

BTW, the 45% speedup I mentioned is only for the force calculation. It gives around 20% speedup for the entire integrator.run in this case.

Reply:
Thanks for explanation. As @govarguz already pointed, could you add the benchmark script to testsuite (if it is feasible to run it on travis)

Reply:
I've added the script as a test (potential_array_test). I also modified it to add new particles of a different type that resizes the potentialArray in between integrator runs.